Minor changes to AffectEntityEvent and ExplosionEvent.Detonate

Yeregorix opened the pull request:
Impl PR : [Sponge#3564](https://github.com/SpongePowered/Sponge/pull/3564)

Changes:
- Make `entities` and `filterAffectedLocations` unmodifiable.
- Make `filterEntities` and `filterAffectedLocations` return void.
- Fix event impl class hierarchy (enforces the presence of `SPAWN_TYPE` context key)
- Exclude `SpawnEntityEvent` from `SpongeEventFactoryTest#testCreate` because the context key is not mocked.

Yeregorix committed:
Make AffectEntityEvent#entities() unmodifiable

gabizou commented on src/main/java/org/spongepowered/api/event/entity/AffectEntityEvent.java:
```diff
@@ -73,6 +71,7 @@
      *
      * @return The List
```
Just clarify what this list is returning, whether it's the original entities, the new entities affected, or the filtered entities that won't be affected.

Yeregorix committed:
Clarify AffectEntityEvent#entities() javadoc

Yeregorix commented on src/main/java/org/spongepowered/api/event/entity/AffectEntityEvent.java:
```diff
@@ -109,17 +98,5 @@
-        }
-        return removedEntities;
-    }
+    List<Entity> filterEntities(Predicate<Entity> predicate);
```
To optimize I suggest returning void. I doubt many plugins need this list. For the few ones that need it, elements can still be collected in the predicate.

dualspiral replied:
I'm with you there I think. I dunno if @gabizou or @Zidane have an opposing view, but in general I think plugins are going to care about what _is_ going to be affected, not what _isn't_. I think we provide the original list of entities that are affected anyway, right? Plugins could also diff that if they need all entities affected by the event regardless.

Yeregorix replied:
There is `List<EntitySnapshot> entitySnapshots()` which is immutable but must be called at `Order#PRE` to be initialized.

Yeregorix commented:
I just saw that [ExplosionEvent.Detonate](https://github.com/SpongePowered/SpongeAPI/blob/api-8/src/main/java/org/spongepowered/api/event/world/ExplosionEvent.java#L99) is in similar situation (`affectedLocations()` and `filterAffectedLocations(Predicate)`). Should I make this list unmodifiable too?

gabizou commented:
> Should I make this list unmodifiable too?

yep!

Yeregorix committed:
Make Detonate#affectedLocations() unmodifiable

Yeregorix commented:
Looks like I have an issue with `event-impl-gen`. I ran `clean` then `genEventImpl` gradle tasks multiple times but the generated event doesn't inherit `AbstractDetonateEvent`.

Test code:
```java
    @Listener
    public void onDetonate(ExplosionEvent.Detonate event) {
        System.out.println(event.getClass().getName());
        System.out.println(event.getClass().getSuperclass().getName());
    }
```

Result:
```
org.spongepowered.api.event.ExplosionEvent$Detonate$Impl
org.spongepowered.api.event.impl.entity.AbstractAffectEntityEvent
```

Same problem for `SpawnEntityEvent`:
```
org.spongepowered.api.event.SpawnEntityEvent$Pre$Impl
org.spongepowered.api.event.impl.entity.AbstractAffectEntityEvent
```

It seems that the first superclass is always ignored despite the `@ImplementedBy` annotation.

Yeregorix commented:
I have probably identified the problem. This sounds like a priority issue with the `ImplementedBy` annotation. 

I can easily fix it by increasing priority of the annotation on the proper interface but I find it counterintuitive that `event-impl-gen` chooses the furthest annotation by default.

https://github.com/SpongePowered/event-impl-gen/blob/6f28baf367d2e95cbdd314cd6160468e2a0209d6/src/main/java/org/spongepowered/eventimplgen/processor/EventImplWriter.java#L184
`priority > max` would make more sense here IMO.

Yeregorix committed:
Fix SpawnEntityEvent and ExplosionEvent.Detonate impl priority

Yeregorix committed:
Exclude SpawnEntityEvent and all sub-events from SpongeEventFactoryTest#testCreate

AbstractSpawnEntityEvent#init requires context key SPAWN_TYPE but the mocked event context doesn't have it

Yeregorix committed:
Make filterEntities and filterAffectedLocations return void

dualspiral commented:
https://github.com/SpongePowered/SpongeAPI/commit/3481b71ed1d7d9f9e5bcf8c55628ff1905da2ad6
